### DISCUSSION TITLE
What should Show do?

### DISCUSSION THREAD

**Discussion by rossabaker** [http4s/http4s#4002] [General]
It came up on https://github.com/http4s/http4s/issues/3998 with @isomarcte and @Lasering.

What are people using `Show` for in http4s?  We've never quite established whether `Show` represents the (HTTP/1) wire rendering of a type, or a source code representation (which is impaired by Cats' regrettable choice of not quoting its `Show[String]`).  That nobody knows has made me eschew the whole thing, but we have them, and people like them, so it would be nice to agree on their purpose.

**Comment by Daenyth** [comment]
I'd want to have the rendering; for example, I should be able to paste `myUri.show` into curl and have it work

**Reply by rossabaker** [reply]
Into curl and not into a REPL?

**Comment by Lasering** [comment]
I think this is appropriate https://github.com/typelevel/cats/pull/3627

**Reply by rossabaker** [reply]
I went back and read it again, and I think I'm mildly :+1: on it.  But we need Cats to move first.

**Comment by isomarcte** [comment]
I'm of the opinion that Show should generally be the same as what the toString generates.

**Reply by rossabaker** [reply]
I think I agree with this.  What should the `toString` generate?

**Reply by isomarcte** [reply]
I don't know that we can give a universal answer to this. On the one hand there is an argument that it should be "Type(member, member, ...)" since that is what the compiler generates for case classes, but on the other hand there are significant deviations from that pattern in the current http4s code base that I don't know we can really change because, for better or worse, people are using them for non-debug code paths. For example, this is a pattern I see often,

```scala
scala> new URI(uri"http://192.0.2.1".toString)
val res0: java.net.URI = http://192.0.2.1
```

To be clear, I'd much rather have people using `new Uri(uri.renderString)` or something more explicit, but unfortunately that's not what always happens.

However, if we were to change to pattern defined by the auto generated toString for case classes, we would likely break many users. That being said, for the purposes of Show/toString I personally think that "Type(member, member, ...)" is much more useful because obviously if I'm debugging and a `org.http4s.Uri` is rendered to a log I can no longer tell if it is really a Uri or a String.

The only way I can see to give a universal answer to this for http4s would be to deprecate all the types which don't align with whatever pattern we choose to use for toString/Show. _Maybe_ that could work since we may have to do that anyway with many types for the cats-uri work...but that would be a significant change.

**Reply by rossabaker** [reply]
Yeah, we're in an awkward place because we haven't been consistent, and making it consistent breaks people who are doing things they shouldn't have been doing but lots of people do anyway.

I have not done a rigorous analysis to see the current state of things.
